drivers/sensor/hdc20x0/hdc20x0.c

```diff
@@ -0,0 +1,243 @@
+{
+	if (hdc20x0_read_temperature(data, config)) {
+		return -EIO;
+	}
+	if (hdc20x0_read_humidity(data, config)) {
+		return -EIO;
+	}
+	return 0;
+}
+
+static int hdc20x0_init(const struct device *dev)
+{
+	struct hdc20x0_data *data = dev->data;
+	const struct hdc20x0_cfg *config = dev->config;
+	int ret;
+
+	data->i2c_master = device_get_binding(config->i2c_bus_name);
+	if (data->i2c_master == NULL) {
+		LOG_ERR("Failed to get I2C device");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Init hdc20x0");
+
+	/* Enable Automatic Measurement Mode at 5Hz */
+	if (i2c_reg_write_byte(data->i2c_master, config->i2c_addr,
+			       hdc20x0_REG_RESET_DRDY_INT_CONF, hdc20x0_AMM)) {
+		return -EIO;
+	}
+	/*
+	 * We enable both temp and humidity measurement.
+	 * However the measurement won't start even in AMM until triggered.
+	 */
+	ret = i2c_reg_write_byte(data->i2c_master, config->i2c_addr,
+				 hdc20x0_REG_CONF, hdc20x0_MEAS_TRIG);
+	if (ret) {
+		LOG_ERR("Unable to set up measurement\n");
+		return ret;
+	}
+
+	LOG_DBG("Init OK");
+	return 0;
+}
+
+static int hdc20x0_sample_fetch(const struct device *dev,
+				enum sensor_channel chan)
+{
+	struct hdc20x0_data *data = dev->data;
+	const struct hdc20x0_cfg *config = dev->config;
+
+	LOG_DBG("Fetching sample from hdc20x0");
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		if (hdc20x0_read_temperature(data, config)) {
+			LOG_ERR("Failed to measure temperature");
+			return -EIO;
+		}
+		break;
+	case SENSOR_CHAN_HUMIDITY:
+		if (hdc20x0_read_humidity(data, config)) {
+			LOG_ERR("Failed to measure humidity");
+			return -EIO;
+		}
+		break;
+	case SENSOR_CHAN_ALL:
+		if (hdc20x0_read_all(data, config)) {
+			LOG_ERR("Failed to measure temperature and humidity");
+			return -EIO;
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int hdc20x0_channel_get(const struct device *dev,
+			       enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct hdc20x0_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		val->val1 = data->temp_val1;
+		val->val2 = data->temp_val2;
+		break;
+	case SENSOR_CHAN_HUMIDITY:
+		val->val1 = data->humidity_val1;
+		val->val2 = data->humidity_val2;
+		break;
+	case SENSOR_CHAN_ALL:
+		val->val1 = data->temp_val1;
+		val->val2 = data->temp_val2;
+		(val + 1)->val1 = data->humidity_val1;
+		(val + 1)->val2 = data->humidity_val2;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api hdc20x0_api_funcs = {
+	.sample_fetch = hdc20x0_sample_fetch,
+	.channel_get = hdc20x0_channel_get,
+};
+
+static struct hdc20x0_data hdc20x0_data_0;
+static const struct hdc20x0_cfg hdc20x0_cfg_0 = { .i2c_bus_name =
+							  DT_INST_BUS_LABEL(0),
+						  .i2c_addr =
+							  DT_INST_REG_ADDR(0) };
+
+DEVICE_AND_API_INIT(hdc20x0, DT_INST_LABEL(0), hdc20x0_init, &hdc20x0_data_0,
+		    &hdc20x0_cfg_0, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,
+		    &hdc20x0_api_funcs);
```

Comment:
It would be easy to make this driver multi-instance as described here: https://docs.zephyrproject.org/latest/guides/dts/howtos.html#option-1-create-devices-using-instance-numbers

Please consider doing so.

Reply:
Hi!

I'm a bit of a newbie jumping in here since I also have a need for HDC20x0 series driver.

So if I want to have two of these on the same i2c bus, I change the end of the file to use either DT_INST_FOREACH_STATUS_OKAY(CREATE_MY_DEVICE) type macros to do the same procedures for all instances
or 
instantiate one-by-one
by checking each label and doing instance specific initializations in a hardcoded way?

Thanks and sorry if this is not a good place to ask :)

Reply:
Hello and welcome.

This is not about whether there multiple instances on the same I2C bus, but about whether there are multiple instances in the entire devicetree. If the driver is made multi-instance as requested here, then you as an application author will only need to declare the appropriate nodes in your devicetree. The driver will create the right devices then.

Reply:
Same suggestion. Make it multi-instance. I guess every new driver should be implemnted in that way.